File: crates/wrapper/src/overlay.rs
use std::{
    collections::HashMap,
    fs, io,
    path::{Path, PathBuf},
    sync::OnceLock,
};

use crate::{exe_dir, patch_fmt, upk};
use crate::mod_toml_mini;

static PATCH_MAP: OnceLock<HashMap<String, Vec<patch_fmt::CdoPatch>>> = OnceLock::new();
static CACHE: OnceLock<std::sync::Mutex<HashMap<String, PathBuf>>> = OnceLock::new();

fn patch_map() -> &'static HashMap<String, Vec<patch_fmt::CdoPatch>> {
    PATCH_MAP.get_or_init(|| unsafe { load_mods_dir() })
}
unsafe fn load_mods_dir() -> HashMap<String, Vec<patch_fmt::CdoPatch>> {
    let mut map: HashMap<String, Vec<patch_fmt::CdoPatch>> = HashMap::new();

    let Some(exe) = exe_dir() else { return map };
    let mods_dir = PathBuf::from(&exe).join("Mods");
    let Ok(rd) = fs::read_dir(&mods_dir) else {
        return map;
    };

    for entry in rd.flatten() {
        let path = entry.path();

        if path.is_dir() {
            let toml_path = path.join("mod.toml");
            if toml_path.exists() {
                load_mod_toml(&path, &toml_path, &mut map);
            }
        } else {
            let fname = path.file_name().and_then(|n| n.to_str()).unwrap_or("");
            let pkg = if let Some(r) = fname.strip_prefix("ScriptPatch_") {
                r.strip_suffix(".bin").unwrap_or("")
            } else {
                continue;
            };
            if pkg.is_empty() {
                continue;
            }
            let Ok(data) = fs::read(&path) else { continue };
            if let Ok(patches) = patch_fmt::load_patch_bin(&data) {
                map.entry(pkg.to_ascii_lowercase())
                    .or_default()
                    .extend(patches);
            }
        }
    }

    map
}

fn load_mod_toml(
    mod_dir: &Path,
    toml_path: &Path,
    map: &mut HashMap<String, Vec<patch_fmt::CdoPatch>>,
) {
    let Ok(text) = fs::read_to_string(toml_path) else {
        return;
    };
    let patches = mod_toml_mini::parse(&text, mod_dir);
    for (pkg, cdo_patches) in patches {
        map.entry(pkg).or_default().extend(cdo_patches);
    }
}

pub(super) fn get_patched_path(original_path: &str) -> Option<PathBuf> {
    let stem = Path::new(original_path)
        .file_stem()?
        .to_str()?
        .to_ascii_lowercase();

    let patches = patch_map().get(&stem)?;
    if patches.is_empty() {
        return None;
    }

    let cache = CACHE.get_or_init(|| std::sync::Mutex::new(HashMap::new()));
    let key = original_path.to_ascii_lowercase();

    {
        let c = cache.lock().ok()?;
        if let Some(p) = c.get(&key) {
            if p.exists() {
                return Some(p.clone());
            }
        }
    }

    let raw = fs::read(original_path).ok()?;
    let patched = upk::apply_cdo_patches(&raw, patches).ok()?;

    let tmp_dir = temp_dir();
    fs::create_dir_all(&tmp_dir).ok()?;
    let out_path = tmp_dir.join(format!("{}_{:016x}.upk", stem, fnv1a(&raw)));
    fs::write(&out_path, &patched).ok()?;

    let mut c = cache.lock().ok()?;
    c.insert(key, out_path.clone());
    Some(out_path)
}

fn temp_dir() -> PathBuf {
    std::env::temp_dir().join("ue3mods")
}

fn fnv1a(data: &[u8]) -> u64 {
    let mut h = 0xcbf2_9ce4_8422_2325u64;
    for &b in data {
        h ^= b as u64;
        h = h.wrapping_mul(0x0000_0100_0000_01B3);
    }
    h
}

File: wrapper/src/overlay.cpp
#define WIN32_LEAN_AND_MEAN
#include "overlay.hpp"
#include "log.hpp"
#include "mod_toml_mini.hpp"
#include "patch_fmt.hpp"
#include "upk.hpp"
#include "util.hpp"
#include <cstdio>
#include <filesystem>
#include <fstream>
#include <iterator>
#include <mutex>
#include <unordered_map>
#include <windows.h>

namespace fs = std::filesystem;
using PatchMap = std::unordered_map<std::string, std::vector<CdoPatch>>;

static uint64_t fnv1a(const uint8_t *d, size_t n) {
  uint64_t h = 0xcbf29ce484222325ULL;
  for (size_t i = 0; i < n; ++i) {
    h ^= d[i];
    h *= 0x100000001B3ULL;
  }
  return h;
}
static std::string hex64(uint64_t v) {
  char buf[17];
  snprintf(buf, sizeof(buf), "%016llx", (unsigned long long)v);
  return buf;
}

static std::once_flag s_once;
static PatchMap s_patch_map;
static std::mutex s_cache_mtx;
static std::unordered_map<std::string, std::string> s_cache;

static PatchMap load_mods_dir() {
  PatchMap map;
  std::wstring wdir = get_exe_dir();
  if (wdir.empty())
    return map;
  fs::path mods_dir = fs::path(wdir) / L"Mods";
  try {
    for (const auto &entry : fs::directory_iterator(mods_dir)) {
      const auto &p = entry.path();
      if (entry.is_directory()) {
        auto toml = p / "mod.toml";
        if (!fs::exists(toml))
          continue;
        std::ifstream f(toml);
        if (!f)
          continue;
        std::string text((std::istreambuf_iterator<char>(f)), {});
        auto patches = mod_toml_mini::parse(text, p.string());
        for (auto &[pkg, cdo] : patches) {
          auto &dst = map[pkg];
          dst.insert(dst.end(), cdo.begin(), cdo.end());
        }
      } else {
        std::string fname = p.filename().string();
        const std::string pfx = "ScriptPatch_";
        if (fname.compare(0, pfx.size(), pfx) != 0)
          continue;
        std::string rest = fname.substr(pfx.size());
        if (rest.size() < 4 || rest.substr(rest.size() - 4) != ".bin")
          continue;
        std::string pkg = rest.substr(0, rest.size() - 4);
        if (pkg.empty())
          continue;
        std::ifstream f(p, std::ios::binary);
        if (!f)
          continue;
        std::vector<uint8_t> data((std::istreambuf_iterator<char>(f)), {});
        try {
          auto ps = load_patch_bin(data.data(), data.size());
          std::string key = pkg;
          for (auto &c : key)
            c = (char)tolower((unsigned char)c);
          auto &dst = map[key];
          dst.insert(dst.end(), ps.begin(), ps.end());
        } catch (...) {
        }
      }
    }
  } catch (...) {
  }
  return map;
}

static void ensure_loaded() {
  std::call_once(s_once, []() { s_patch_map = load_mods_dir(); });
}

namespace overlay {

std::string get_patched_path(const std::string &original_path) {
  fs::path p(original_path);
  std::string stem = p.stem().string();
  for (auto &c : stem)
    c = (char)tolower((unsigned char)c);

  ensure_loaded();

  auto it = s_patch_map.find(stem);
  if (it == s_patch_map.end() || it->second.empty())
    return {};

  std::string key = original_path;
  for (auto &c : key)
    c = (char)tolower((unsigned char)c);

  {
    std::lock_guard<std::mutex> lk(s_cache_mtx);
    auto cit = s_cache.find(key);
    if (cit != s_cache.end() && fs::exists(cit->second))
      return cit->second;
  }

  std::ifstream f(original_path, std::ios::binary);
  if (!f)
    return {};
  std::vector<uint8_t> raw((std::istreambuf_iterator<char>(f)), {});

  std::vector<uint8_t> patched;
  try {
    patched = apply_cdo_patches(raw.data(), raw.size(), it->second);
  } catch (const std::exception &ex) {
    log_err("apply_cdo_patches: %s", ex.what());
    return {};
  }

  auto tmp = fs::temp_directory_path() / "ue3mods";
  try {
    fs::create_directories(tmp);
  } catch (...) {
    return {};
  }

  auto out_path =
      tmp / (stem + "_" + hex64(fnv1a(raw.data(), raw.size())) + ".upk");
  {
    std::ofstream of(out_path, std::ios::binary);
    if (!of)
      return {};
    of.write(reinterpret_cast<const char *>(patched.data()), patched.size());
  }

  std::string result = out_path.string();
  {
    std::lock_guard<std::mutex> lk(s_cache_mtx);
    s_cache[key] = result;
  }
  return result;
}

} // namespace overlay
